### DISCUSSION TITLE
Library integration for ESP32

### DISCUSSION THREAD

**Discussion by engel2** [Open-Agriculture/AgIsoStack-plus-plus#225] [Q&A]
Hi guys,
I'm really enthusiastic about your work with the isobus-plus-plus project. 
I got the project compiled in linux and windows and also got the examples working with the PCAN driver, finally seeing the CAN messages on the bus. :)
Now I tried to integrate the library to a ESP32 platform using PlatformIO, but struggeling. Do you have any hints how to get it working? I added the library to the lib_deps by using this in the platformio.ini:
lib_deps =
    https://github.com/ad3154/Isobus-plus-plus.git

Now there are several errors, some of seem to be caused that C++14 is necessary?

Maybe a short tutourial would be nice, how to integrate it for "dummies". Let me know, if I should note it step by step so it could be published in the readthedocs-section.

Do you have any hints for me? 

Thanks a lot for your great work!

**Comment by GwnDaan** [comment]
Hi! Thanks for your kind words. I'm currently running a small personal ESP32 project that successfully integrates isobus++. If you have the time to make a step-by-step guide on how to implement isobus-plus-plus with ESP on PlatformIO, that would be greatly appreciated!

Regarding the issue you are facing, I also faced it a while back if I remember correctly. But I'm unsure how I fixed it back then. My current `platformio.ini` looks like this, maybe that it helps resolving your issue?:
```ini
[env:debug]
platform = espressif32
board = esp32-s3-custom
framework = espidf

build_type = debug
upload_protocol = esptool
debug_tool = esp-builtin
board_build.embed_txtfiles = src/vtpooldata.iop

monitor_dtr = 0
monitor_speed = 115200
monitor_filters = esp32_exception_decoder

lib_deps = https://github.com/ad3154/Isobus-plus-plus.git
```
I'm using the `ESP-IDF v5` framework here, instead of the (more popular) arduino framework. Oh and note that `esp32-s3-custom` is my own board file, so use yours here. 

C++14 is indeed a requirement for integrating this library. There should be an option somewhere in PlatformIO to enforce it. Let me know if you are able to fix it. Otherwise the logs of building and/or steps to reproduce would be useful to help you solve this problem.

Thanks for your interest in the project!

**Reply by GwnDaan** [reply]
Oh also, the `.vscode/c_cpp_properties.json` file generated by PlatformIO indicates which C++ standard is being used. For me it indicates:
```           json
"cStandard": "c17",
"cppStandard": "c++20",
```

Note that it is an auto generated file, and should therefore only be used as indication:
```
// !!! WARNING !!! AUTO-GENERATED FILE!
// PLEASE DO NOT MODIFY IT AND USE "platformio.ini":
// https://docs.platformio.org/page/projectconf/section_env_build.html#build-flags
```

**Reply by bjoernpoettker** [reply]
My `.vscode/c_cpp_properties.json` contains the following lines:

 ```
 "cStandard": "c99",
 "cppStandard": "c++14",
```

**Reply by engel2** [reply]
I think 
`"cStandard": "c17",
            "cppStandard": "c++20",`
seems to be default after installing VSCode. This options works in my setup.

**Reply by bjoernpoettker** [reply]
Hi @GwnDaan

You embed the iop-file in the platformio config - how did you load it into the VirtualTerminal?

**Reply by GwnDaan** [reply]
> Hi @GwnDaan
> 
> You embed the iop-file in the platformio config - how did you load it into the VirtualTerminal? 

Hey! Coincidentally, I explained this a few days ago here: https://github.com/ad3154/Isobus-plus-plus/discussions/229#discussioncomment-5498602

**Comment by bjoernpoettker** [comment]
Hi!
I'm using the library on a Olimex ESP32-EVB with PlatformIO.

I added the folowing line in the MakeList.txt of the root folder:

```
cmake_minimum_required(VERSION 3.16.0)

set(CMAKE_CXX_STANDARD 14)
include($ENV{IDF_PATH}/tools/cmake/project.cmake)
project(ISOBUS)
```

**Comment by engel2** [comment]
Thanks a lot for you quick support, guys!

Right now I was able to comile the project and honestly I don't know what I made wrong the last hours.

The **only** thing I changed right now after creating a new project in PlatformIO with the ESP-IDF framework was adding the library, so my platformio.ini looks like:
```
[env:esp32-evb]
platform = espressif32
board = esp32-evb
framework = espidf

lib_deps = https://github.com/ad3154/Isobus-plus-plus.git
```
Maybe it's important to disable the Espressif IDF extension. I will try to understand this with a clean installation of VSCode and all the necessary extensions.

**Comment by joaooliveira82** [comment]
Hi guys,
I've been trying to compile this library for days using VS Code + PlatformIO + esp IDF for esp32 but many compilation errors appear, anyone who could help?

This is the terminal output:
``` 
Successfully created esp32 image.
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_NAME_filter.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_address_claim_state_machine.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_callbacks.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_control_function.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_extended_transport_protocol.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_identifier.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_internal_control_function.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_message.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_message_frame.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_network_configuration.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_network_manager.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_parameter_group_number_request_protocol.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_partnered_control_function.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_protocol.o
Compiling .pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_stack_logger.o
In file included from /home/joao/.platformio/packages/toolchain-xtensa-esp32/xtensa-esp32-elf/include/c++/11.2.0/memory:77,
                 from .pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/include/isobus/isobus/can_control_function.hpp:16,
                 from .pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/include/isobus/isobus/can_extended_transport_protocol.hpp:15,
                 from .pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/src/can_extended_transport_protocol.cpp:10:
/home/joao/.platformio/packages/toolchain-xtensa-esp32/xtensa-esp32-elf/include/c++/11.2.0/bits/shared_ptr.h: In instantiation of 'std::shared_ptr<_Tp> std::dynamic_pointer_cast(std::shared_ptr<_Tp>&&) [with _Tp = isobus::InternalControlFunction; _Up = isobus::ControlFunction]':
.pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/src/can_extended_transport_protocol.cpp:113:160:   required from here
/home/joao/.platformio/packages/toolchain-xtensa-esp32/xtensa-esp32-elf/include/c++/11.2.0/bits/shared_ptr.h:648:23: error: 'dynamic_cast' not permitted with '-fno-rtti'
  648 |       if (auto* __p = dynamic_cast<typename _Sp::element_type*>(__r.get()))
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/src/can_internal_control_function.cpp: In static member function 'static std::shared_ptr<isobus::InternalControlFunction> isobus::InternalControlFunction::create(isobus::NAME, uint8_t, uint8_t)':
.pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/src/can_internal_control_function.cpp:33:74: error: 'constexpr isobus::CANLibBadge<T>::CANLibBadge() [with T = isobus::ControlFunction]' is private within this context
   33 |                 CANNetworkManager::CANNetwork.on_control_function_created(controlFunction, {});
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
In file included from .pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/include/isobus/isobus/can_internal_control_function.hpp:16,
                 from .pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/src/can_internal_control_function.cpp:11:
.pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/include/isobus/isobus/can_badge.hpp:34:17: note: declared private here
   34 |                 CANLibBadge() = default;
      |                 ^~~~~~~~~~~
*** [.pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_extended_transport_protocol.o] Error 1
*** [.pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_internal_control_function.o] Error 1
.pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/src/can_partnered_control_function.cpp: In static member function 'static std::shared_ptr<isobus::PartneredControlFunction> isobus::PartneredControlFunction::create(uint8_t, std::vector<isobus::NAMEFilter>)':
.pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/src/can_partnered_control_function.cpp:32:74: error: 'constexpr isobus::CANLibBadge<T>::CANLibBadge() [with T = isobus::ControlFunction]' is private within this context
   32 |                 CANNetworkManager::CANNetwork.on_control_function_created(controlFunction, {});
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
In file included from .pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/include/isobus/isobus/can_partnered_control_function.hpp:17,
                 from .pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/src/can_partnered_control_function.cpp:11:
.pio/libdeps/esp32doit-devkit-v1/AgIsoStack__/isobus/include/isobus/isobus/can_badge.hpp:34:17: note: declared private here
   34 |                 CANLibBadge() = default;
      |                 ^~~~~~~~~~~
*** [.pio/build/esp32doit-devkit-v1/lib083/AgIsoStack__/isobus/src/can_partnered_control_function.o] Error 1
=========================================================== [FAILED] Took 29.05 seconds ===========================================================

 *  The terminal process "platformio 'run', '--environment', 'esp32doit-devkit-v1'" terminated with exit code: 1. 
 *  Terminal will be reused by tasks, press any key to close it. 

```
Thanks a lot for your work!

**Reply by ad3154** [reply]
Hello there! 

I think maybe there's 2 different things going on in this one. One of them I can at least say is our problem, which should be getting fixed in #303 and is a strange issue we saw when compiling for ESP platforms only. The other seems to be that you either have `-fno-rtti` set as a compiler flag, or that's the default in which case you might need to add `-frtti` to enable runtime type information?

If you want to try with our fixes for `isobus::CANLibBadge<T>::CANLibBadge() [with T = isobus::ControlFunction]' is private within this context` you can try to switch to the `esp-cf-constructor-visibility` branch:

```
git pull
git switch esp-cf-constructor-visibility
```

**Reply by ad3154** [reply]
Update: This fix is now in `main` so `git pull` should be sufficient to get it if you are on the main branch.

**Reply by Schine73** [reply]
Hi!
I'm new to this library and i am also trying to compile it using VSCode and PlatformIO with espIDF and I am getting the same errors as @joaooliveira82.

I was able to resolve the issue with the `-fno-rtti` flag being set by adding `build_unflags = -fno-rtti` to the platformio.ini file.

However I am still getting the second error: `'constexpr isobus::CANLibBadge<T>::CANLibBadge() [with T = isobus::ControlFunction]' is private within this context`.

**Reply by ad3154** [reply]
Thanks for reporting this. I was hoping this would be fixed earlier, but It seems like it's actually mad that we're trying to use `CanLibBadge<ControlFunction>` from `InternalControlFunction` and likewise probably would be mad about the same thing in `PartneredControlFunction` despite inheriting from `ControlFunction`, and which works on other compilers... 😣

An easy fix might be for us to add overloads of `on_control_function_created` that take either one instead of `CanLibBadge<ControlFunction>`, or perhaps to make the badge's `T` member protected instead of private.

I will open an issue for this and get it fixed on our end ASAP

**Reply by ad3154** [reply]
Created issue #322

**Reply by ad3154** [reply]
OK @Schine73 I believe we have fixed this in #323 so updating the library (which platform IO might do automatically?) should fix the compilation issue you and @joaooliveira82  were seeing.

**Comment by Ditzebecker** [comment]
Hello together, I'm not sure if my issue is on the right spot here. I also tried to get the library running with an ESP32 and the virtual terminal example. After some challenges that I faced (like there was no "create" function in the can_network_manager.hpp anymore (seems liks it's replaced by "create_internal_control_function", I was able to compile everything and flash it on the esp32. 

I now attached the esp32 to a CLAAS Communicator II. Unfortunately the object pool is not loading. On the serial monitor I get following output: 

```
rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:7172
load:0x40078000,len:15540
load:0x40080400,len:4
ho 8 tail 4 room 4
load:0x40080404,len:3904
entry 0x40080638
I (31) boot: ESP-IDF 5.2.1 2nd stage bootloader
I (31) boot: compile time Jun  9 2024 00:15:29
I (31) boot: Multicore bootloader
I (35) boot: chip revision: v3.1
I (39) boot.esp32: SPI Speed      : 40MHz
I (44) boot.esp32: SPI Mode       : DIO
I (48) boot.esp32: SPI Flash Size : 4MB
I (53) boot: Enabling RNG early entropy source...
I (58) boot: Partition Table:
I (62) boot: ## Label            Usage          Type ST Offset   Length
I (69) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (76) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (84) boot:  2 factory          factory app      00 00 00010000 00100000
I (91) boot: End of partition table
I (96) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=47ce4h (294116) map
I (205) esp_image: segment 1: paddr=00057d0c vaddr=3ffb0000 size=02398h (  9112) load
I (208) esp_image: segment 2: paddr=0005a0ac vaddr=40080000 size=05f6ch ( 24428) load
I (220) esp_image: segment 3: paddr=00060020 vaddr=400d0020 size=647dch (411612) map
I (362) esp_image: segment 4: paddr=000c4804 vaddr=40085f6c size=07be8h ( 31720) load
I (381) boot: Loaded app from partition at offset 0x10000
I (381) boot: Disabling RNG early entropy source...
I (393) cpu_start: Multicore app
I (401) cpu_start: Pro cpu start user code
I (402) cpu_start: cpu freq: 160000000 Hz
I (402) cpu_start: Application information:
I (405) cpu_start: Project name:     TestAgIsoStack
I (410) cpu_start: App version:      1
I (415) cpu_start: Compile time:     Jun  9 2024 00:14:27
I (421) cpu_start: ELF file SHA256:  c4c69b20c...
I (426) cpu_start: ESP-IDF:          5.2.1
I (431) cpu_start: Min chip rev:     v0.0
I (435) cpu_start: Max chip rev:     v3.99 
I (440) cpu_start: Chi
 I (1860) AgIsoStack: [Info][NM]: Partnered control function with name a0001d000d142682 has claimed address 38 on channel 0.
W (1916) AgIsoStack: [Warn][VT]: Server sent function '4610535980821053440' which we do not support
W (1916) AgIsoStack: [Warn][VT]: Server sent function '4610535980821053440' which we do not support
Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled.

Core  0 register dump:
PC      : 0x400d416c  PS      : 0x00060a30  A0      : 0x800d4190  A1      : 0x3ffbdaa0  
A2      : 0x00000001  A3      : 0x3ffbe657  A4      : 0x00000005  A5      : 0x3ffbdc04  
A6      : 0x3ffbdd3c  A7      : 0x3ffbdbf0  A8      : 0x000000f1  A9      : 0x3ffbdd3c  
A10     : 0x60000000  A11     : 0x00000003  A12     : 0x00060a20  A13     : 0x0000000c  
A14     : 0x00000074  A15     : 0x00000010  SAR     : 0x00000018  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x000000f1  LBEG    : 0x4000c349  LEND    : 0x4000c36b  LCOUNT  : 0xffffffff  


Backtrace: 0x400d4169:0x3ffbdaa0 0x400d418d:0x3ffbdbc0 0x400d643e:0x3ffbdbe0 0x400d6621:0x3ffbdc30 0x400dd799:0x3ffbdcb0 0x400e2bb9:0x3ffbdcf0 0x400d566d:0x3ffbe6d0 0x400d7690:0x3ffbe790 0x400d8897:0x3ffbe7e0 0x400d890a:0x3ffbe810 0x400d43f6:0x3ffbe830 0x400d446d:0x3ffbe870 0x4012f7d5:0x3ffbe8b0 0x400f89f1:0x3ffbe8d0 0x400f49b8:0x3ffbe900 0x400872c5:0x3ffbe920
  #0  0x400d4169 in isobus::CANHardwareInterface::transmit_can_frame(isobus::CANMessageFrame const&) at .pio/libdeps/esp32dev/AgIsoStack__/hardware_integration/src/can_hardware_interface.cpp:317 (discriminator 2)
  #1  0x400d418d in isobus::send_can_message_frame_to_hardware(isobus::CANMessageFrame const&) at .pio/libdeps/esp32dev/AgIsoStack__/hardware_integration/src/can_hardware_interface.cpp:158
  #2  0x400d643e in isobus::CANNetworkManager::send_can_message_raw(unsigned long, unsigned char, unsigned char, unsigned long, unsigned char, void const*, unsigned long) const at .pio/libdeps/esp32dev/AgIsoStack__/isobus/src/can_network_manager.cpp:1085
  #3  0x400d6621 in isobus::CANNetworkManager::send_can_message(unsigned long, unsigned char const*, unsigned long, std::shared_ptr<isobus::InternalControlFunction>, std::shared_ptr<isobus::ControlFunction>, isobus::CANIdentifier::CANPriority, void (*)(unsigned long, unsigned long, std::shared_ptr<isobus::InternalControlFunction>, std::shared_ptr<isobus::ControlFunction>, bool, void*), void*, bool (*)(unsigned long, unsigned long, unsigned long, unsigned char*, void*)) at .pio/libdeps/esp32dev/AgIsoStack__/isobus/src/can_network_manager.cpp:207 (discriminator 6)
  #4  0x400dd799 in isobus::VirtualTerminalClient::send_message_to_vt(unsigned char const*, unsigned long, isobus::CANIdentifier::CANPriority) const at .pio/libdeps/esp32dev/AgIsoStack__/isobus/src/isobus_virtual_terminal_client.cpp:1747 (discriminator 2)
  #5  0x400e2bb9 in isobus::VirtualTerminalClient::process_rx_message(isobus::CANMessage const&, void*) at .pio/libdeps/esp32dev/AgIsoStack__/isobus/src/isobus_virtual_terminal_client.cpp:3190 (discriminator 1)
  #6  0x400d566d in isobus::CANNetworkManager::process_can_message_for_global_and_partner_callbacks(isobus::CANMessage const&) at .pio/libdeps/esp32dev/AgIsoStack__/isobus/src/can_network_manager.cpp:980 (discriminator 4)
  #7  0x400d7690 in isobus::CANNetworkManager::process_rx_messages() at .pio/libdeps/esp32dev/AgIsoStack__/isobus/src/can_network_manager.cpp:1029
  #8  0x400d8897 in isobus::CANNetworkManager::update() at .pio/libdeps/esp32dev/AgIsoStack__/isobus/src/can_network_manager.cpp:233
  #9  0x400d890a in isobus::periodic_update_from_hardware() at .pio/libdeps/esp32dev/AgIsoStack__/isobus/src/can_network_manager.cpp:294
  #10 0x400d43f6 in isobus::CANHardwareInterface::update() at .pio/libdeps/esp32dev/AgIsoStack__/hardware_integration/src/can_hardware_interface.cpp:382
  #11 0x400d446d in isobus::CANHardwareInterface::update_thread_function() at .pio/libdeps/esp32dev/AgIsoStack__/hardware_integration/src/can_hardware_interface.cpp:420
  #12 0x4012f7d5 in void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) at /home/michael/.platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/13.2.0/bits/invoke.h:61
      (inlined by) std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) at /home/michael/.platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/13.2.0/bits/invoke.h:96
      (inlined by) void std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0u>(std::_Index_tuple<0u>) at /home/michael/.platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/13.2.0/bits/std_thread.h:292
      (inlined by) std::thread::_Invoker<std::tuple<void (*)()> >::operator()() at /home/michael/.platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/13.2.0/bits/std_thread.h:299
      (inlined by) std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() at /home/michael/.platformio/packages/toolchain-xtensa-esp-elf/xtensa-esp-elf/include/c++/13.2.0/bits/std_thread.h:244
  #13 0x400f89f1 in execute_native_thread_routine at /builds/idf/crosstool-NG/.build/xtensa-esp-elf/src/gcc/libstdc++-v3/src/c++11/thread.cc:104 (discriminator 1)
  #14 0x400f49b8 in pthread_task_func at /home/michael/.platformio/packages/framework-espidf/components/pthread/pthread.c:196
  #15 0x400872c5 in vPortTaskWrapper at /home/michael/.platformio/packages/framework-espidf/components/freertos/FreeRTOS-Kernel/portable/xtensa/port.c:134


ELF file SHA256: c4c69b20c

Rebooting...
ets Jul 29 2019 12:21:46
```

... it seems like address claiming is happening (?) but there is function sent by the terminal which doesn't work with the stack? I am kinda new to that ISOBUS area. Is there something which I miss/don't understand? Since it' working fine with my raspberry and the same VT. 
Thanks in advance!

**Reply by GwnDaan** [reply]
Hi there, I'm happy to try to help you out here.

**TLDR**: make sure you have fully applied step 8 of the [ESP32 tutorial](https://agisostack-plus-plus.readthedocs.io/en/latest/Tutorials/ESP32%20PlatformIO.html#id3). Especially the increase of the default thread stack size.

The long answer: since the partnered control function has been claimed by the VT server, I think you're CAN communication is working correctly. Though I've not seen any unsupported function being sent before, so that's new, though I doubt that's the only reason the ESP crashed. Looking at the backtrace that was dumped, it looks like it crashed on the following line: (assuming you were on the latest commit of the main branch here)
https://github.com/Open-Agriculture/AgIsoStack-plus-plus/blob/e069b7c4ece90635954e41d63f4e5802ce99c2ec/hardware_integration/src/can_hardware_interface.cpp#L317

The interesting part here is that `frameHandler` has been checked to be not `nullptr` right before this line. So my bet would be the frame handler pointer got overwritten with other memory because the stack reached it's limit. Hence my suggestion to look at increasing the thread's stack size. If that doesn't solve the issue, we'll need some more diagnostic information like sdkconfig, cantrace, hardware details, etc., to be able to help you out.

Also, the "create" function for an internal control function is already using the new `create_internal_control_function` function in the [esp32 virtual terminal example](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/tree/main/examples/virtual_terminal/esp32_platformio_object_pool) as far as I can tell? Though I'm happy if proven wrong here :)

Lastly, you've found some some small formatting issue here with the logged unsupported function codes, I proposed a fix for it in https://github.com/Open-Agriculture/AgIsoStack-plus-plus/pull/475 :). It would be appreciated if you can try to run the code again once the PR is merged to main and see which function-code the server tries to send. I'm curious to see what the VT wants hahah. I assume you also encounter this log messages when running the example on the raspberry pi?

**Reply by Ditzebecker** [reply]
Hey GwanDaan, thanks for your fast reply and your help!!

I checked step 8 again and unfortunately didn't make it run in the first place (check beyond for new errors..). Few more information and tests from my side (as said before, I am at the start or working with C++ and those things). I writing down in the order in tried. 

**Create function:** 
I was just trying again the tutorial, also with the given example. (created a new project in platformIO and added the github as lib-dep again new, so should be the newest. Also tried pio lib update / pio pkg update, to be sure). When coping this code from tutorial step 7, VS Code throws me the error that the function is not a member of InternalControlFunction .. 

![image](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/assets/60115385/44d9cf7a-6dcc-4fe4-90d4-9215211b5f59)

So I was replacing it by the code from the example folder in the stack with the lines:
```
	auto TestInternalECU = isobus::CANNetworkManager::CANNetwork.create_internal_control_function(TestDeviceNAME, 0, 0x81);
```

**Frame Handler:**
For better debugging, I was creating some lines right before of Line 317 :

```
LOG_INFO("[CANHardwareInterface]: transmit_can_frame called");

if (channel == nullptr) {
    LOG_ERROR("[CANHardwareInterface]: channel is null");
    return false;
}

if (channel->frameHandler == nullptr) {
    LOG_ERROR("[CANHardwareInterface]: frameHandler is null");
    return false;
}

LOG_INFO("[CANHardwareInterface]: channel and frameHandler are valid");

if (channel->frameHandler->get_is_valid()) {
    LOG_INFO("[CANHardwareInterface]: frameHandler is valid, pushing frame to queue");
    channel->messagesToBeTransmittedQueue.push(frame);
#if !defined CAN_STACK_DISABLE_THREADS && !defined ARDUINO
    LOG_INFO("[CANHardwareInterface]: Notifying all threads");
    updateThreadWakeupCondition.notify_all();
#endif
    return true;
} else {
    LOG_WARNING("[CANHardwareInterface]: frameHandler is not valid");
}

LOG_INFO("[CANHardwareInterface]: Exiting transmit_can_frame");
return false;

```
`Update till here: `
In the simple example without VT functions / Client it worked so far and no error were given. The VT was recognizing the ESP32 as 0x81 (as given above). Next step: let's add the VT functions again:

Using the example from the [example main.cpp](examples/virtual_terminal/esp32_platformio_object_pool/src/main.cpp) again. (create_internal_control_function already there. Adding preferred number 0x81)

Attention: The lines from the tutorial in step 7 ... 
```
#include "esp_log.h"
#include "freertos/task.h"
```
where not given in the [example main.cpp](examples/virtual_terminal/esp32_platformio_object_pool/src/main.cpp), therefore the code wasn't working out of the box. 

**The good news**: I don't know what I did different now, but at least ESP32 is not rebooting anymore. That's the log now (I already adjusted the isobus_virtual_terminal_client.cpp lines to %hu and removed the additional debugging lines again for better overview):

```
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:7172
load:0x40078000,len:15540
load:0x40080400,len:4
ho 8 tail 4 room 4
load:0x40080404,len:3904
entry 0x40080638
I (31) boot: ESP-IDF 5.2.1 2nd stage bootloader
I (31) boot: compile time Jun 11 2024 23:59:49
I (31) boot: Multicore bootloader
I (35) boot: chip revision: v3.1
I (39) boot.esp32: SPI Speed      : 40MHz
I (44) boot.esp32: SPI Mode       : DIO
I (48) boot.esp32: SPI Flash Size : 4MB
I (53) boot: Enabling RNG early entropy source...
I (58) boot: Partition Table:
I (62) boot: ## Label            Usage          Type ST Offset   Length
I (69) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (76) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (84) boot:  2 factory          factory app      00 00 00010000 00100000
I (91) boot: End of partition table
I (96) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=48540h (296256) map
I (206) esp_image: segment 1: paddr=00058568 vaddr=3ffb0000 size=02384h (  9092) load
I (209) esp_image: segment 2: paddr=0005a8f4 vaddr=40080000 size=05724h ( 22308) load
I (220) esp_image: segment 3: paddr=00060020 vaddr=400d0020 size=63ee0h (409312) map
I (361) esp_image: segment 4: paddr=000c3f08 vaddr=40085724 size=075dch ( 30172) load
I (379) boot: Loaded app from partition at offset 0x10000
I (379) boot: Disabling RNG early entropy source...
I (391) cpu_start: Multicore app
I (400) cpu_start: Pro cpu start user code
I (400) cpu_start: cpu freq: 160000000 Hz
I (400) cpu_start: Application information:
I (403) cpu_start: Project name:     AgIsoStack
I (408) cpu_start: App version:      1
I (412) cpu_start: Compile time:     Jun 11 2024 23:58:56
I (418) cpu_start: ELF file SHA256:  c24a5c293...
I (424) cpu_start: ESP-IDF:          5.2.1
I (429) cpu_start: Min chip rev:     v0.0
I (433) cpu_start: Max chip rev:     v3.99 
I (438) cpu_start: Chip rev:         v3.1
I (443) heap_init: Initializing. RAM available for dynamic allocation:
I (450) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (456) heap_init: At 3FFB66D8 len 00029928 (166 KiB): DRAM
I (462) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (469) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (475) heap_init: At 4008CD00 len 00013300 (76 KiB): IRAM
I (483) spi_flash: detected chip: generic
I (486) spi_flash: flash io: dio
I (496) main_task: Started on CPU0
I (500) main_task: Calling app_main()
I (500) gpio: GPIO[22]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (504) gpio: GPIO[21]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
[Info][NM]: Partnered control function with name a0001d000d142682 has claimed address 38 on channel 0.
[Warn][VT]: Server sent function '33' which we do not support
[Warn][VT]: Server sent function '33' which we do not support
[Info][VT]: No version label from the VT matched. Client will upload the pool and store it instead.
[Warn][VT]: Server sent function '33' which we do not support
[Warn][VT]: Server sent function '33' which we do not support
[Error][ETP]: Timeout tx session for 0x0E700 (expected CTS)
[Error][VT]: An object pool failed to upload. Resetting connection to VT.
```
Do you have any idea about how to fix these errors?

 Thanks for your help in advance!

**Reply by GwnDaan** [reply]
Hi @Ditzebecker, sorry for the delay. 

The `Server sent function '33' which we do not support` line means it receives messages that contain the status of the auxiliary input device with the AUX-O (auxiliary type 1 objects) standard. But we currently only have support built-in for AUX-N (auxiliary type 2 objects). But as it will not affect the simple VT example, we can discard these warnings for now.

The `Timeout tx session for 0x0E700 (expected CTS)` usually tells us that there are some packets that got lost on the bus, which can be both incoming or outgoing. And since you said:
> it' working fine with my raspberry and the same VT

I'd suggest to make sure you're wiring is correct and follows the canbus specifications

**Reply by Ditzebecker** [reply]
Hey @GwnDaan , now the delay is on my side .. harvest season started .. 

I additionally added following lines for the TWAI Timing and Driver setup: 
```
	twaiTiming.quanta_resolution_hz = 4000000;       // Baudrate Prescaler
  	twaiTiming.tseg_1 = 12;    // PROP_SEG (5) + PHASE_SEG1 (5)
  	twaiTiming.tseg_2 = 3;     // PHASE_SEG2
  	twaiTiming.sjw = 2;       // Synchronization Jump Width 
```
I got it by the ISOBUS Norm, hope I understood it the correct way. 
However, it seems to work now, object pool is now loading! Thanks a lot for your help! 

Best regards!
